File: PolyScript/examples/backup_tool/backup_tool.py
#!/usr/bin/env python3
"""
Example: Backup Tool using PolyScript Click Framework

This demonstrates how the PolyScript framework eliminates boilerplate.
Compare this to the 400+ line manual implementation in the migration guide.

Author: Mathew Burkitt, Dimension Technologies <mathew.burkitt@ditech.ai>
"""

import os
import shutil
from pathlib import Path

# This would import from the framework
# from polyscript_click import polyscript_tool

# For demonstration purposes, we'll show what the code would look like
# when the framework is available:

"""
@polyscript_tool
class BackupTool:
    '''PolyScript-compliant file backup tool with zero boilerplate'''
    
    def add_arguments(self, cmd):
        '''Define tool-specific arguments - applied to all modes automatically'''
        cmd.argument('source', help='Source directory to backup')
        cmd.argument('dest', help='Destination directory')
        cmd.option('--overwrite', is_flag=True, help='Overwrite existing destination')
    
    def status(self):
        '''Show backup status - ONLY business logic, no boilerplate'''
        source = Path(self.ctx.current_args['source'])
        dest = Path(self.ctx.current_args['dest'])
        
        return {
            "source": {
                "path": str(source),
                "exists": source.exists(),
                "size_bytes": self._get_dir_size(source) if source.exists() else 0,
                "file_count": self._get_file_count(source) if source.exists() else 0
            },
            "destination": {
                "path": str(dest),
                "exists": dest.exists(),
                "would_overwrite": dest.exists() and not self.ctx.current_args.get('overwrite', False)
            },
            "backup_needed": source.exists() and (not dest.exists() or self.ctx.current_args.get('overwrite', False))
        }
    
    def test(self):
        '''Dry-run mode - show what would be backed up'''
        source = Path(self.ctx.current_args['source'])
        dest = Path(self.ctx.current_args['dest'])
        
        if not source.exists():
            self.ctx.output("Source directory does not exist", error=True)
            return None  # Framework converts this to exit code 1
        
        return {
            "planned_operations": [{
                "operation": "backup",
                "source": str(source),
                "destination": str(dest),
                "file_count": self._get_file_count(source),
                "size_bytes": self._get_dir_size(source),
                "would_overwrite": dest.exists()
            }],
            "note": "No changes made in test mode"
        }
    
    def sandbox(self):
        '''Test environment and dependencies'''
        source = Path(self.ctx.current_args['source'])
        dest = Path(self.ctx.current_args['dest'])
        
        tests = {
            "source_readable": self._test_readable(source),
            "destination_writable": self._test_writable(dest.parent),
            "sufficient_space": self._test_disk_space(source, dest.parent)
        }
        
        return {
            "dependency_tests": tests,
            "all_passed": all(test == "passed" for test in tests.values())
        }
    
    def live(self):
        '''Perform actual backup'''
        source = Path(self.ctx.current_args['source'])
        dest = Path(self.ctx.current_args['dest'])
        
        if not source.exists():
            self.ctx.output("Source directory does not exist", error=True)
            return None
        
        # Framework handles confirmation prompts automatically based on --force flag
        if dest.exists() and not self.ctx.current_args.get('overwrite', False):
            if not self.ctx.confirm(f"Destination {dest} exists. Overwrite?"):
                return {"status": "cancelled"}
        
        # Actual backup operation - just business logic
        if dest.exists():
            shutil.rmtree(dest)
        
        shutil.copytree(source, dest)
        
        return {
            "operation": "backup_completed",
            "source": str(source),
            "destination": str(dest),
            "files_copied": self._get_file_count(dest),
            "bytes_copied": self._get_dir_size(dest)
        }
    
    # Helper methods - pure business logic
    def _get_dir_size(self, path: Path) -> int:
        if not path.exists():
            return 0
        total = 0
        try:
            for item in path.rglob("*"):
                if item.is_file():
                    total += item.stat().st_size
        except (OSError, PermissionError):
            pass
        return total
    
    def _get_file_count(self, path: Path) -> int:
        if not path.exists():
            return 0
        count = 0
        try:
            for item in path.rglob("*"):
                if item.is_file():
                    count += 1
        except (OSError, PermissionError):
            pass
        return count
    
    def _test_readable(self, path: Path) -> str:
        try:
            return "passed" if path.exists() and os.access(path, os.R_OK) else "failed"
        except Exception:
            return "error"
    
    def _test_writable(self, path: Path) -> str:
        try:
            return "passed" if path.exists() and os.access(path, os.W_OK) else "failed"
        except Exception:
            return "error"
    
    def _test_disk_space(self, source: Path, dest_parent: Path) -> str:
        try:
            if not source.exists() or not dest_parent.exists():
                return "unknown"
            source_size = self._get_dir_size(source)
            free_space = shutil.disk_usage(dest_parent).free
            return "passed" if free_space > source_size * 1.1 else "failed"
        except Exception:
            return "error"


# Usage would be simply:
if __name__ == "__main__":
    BackupTool.run()

# The framework automatically provides:
# - All CLI argument parsing and validation
# - --mode status|test|sandbox|live routing  
# - --json, --verbose, --force standard flags
# - PolyScript v1.0 JSON output formatting
# - Error handling and exit codes
# - Help text generation
# - Confirmation prompts

# Command examples that would work:
# python backup_tool.py status /source /dest
# python backup_tool.py test /source /dest --overwrite
# python backup_tool.py sandbox /source /dest --json
# python backup_tool.py live /source /dest --force --verbose
"""

print("""
POLYSCRIPT CLICK FRAMEWORK EXAMPLE
==================================

This file shows what a backup tool would look like using the PolyScript Click framework.

Key Benefits:
- ZERO boilerplate code
- Developer writes ONLY business logic
- Framework handles all PolyScript compliance automatically
- Automatic CLI generation with Click
- Standard --json, --verbose, --force flags
- PolyScript v1.0 JSON output format
- Error handling and exit codes

Compare this ~100 lines of pure business logic to the 400+ lines of 
boilerplate required in the manual PolyScript implementation.

The framework eliminates:
- Manual argument parsing (Click handles this)
- Mode routing logic (automatic)
- JSON output formatting (automatic)
- Standard flag handling (automatic)
- Error handling boilerplate (automatic)
- Help text generation (automatic)

To use this framework:
1. pip install click
2. Import: from polyscript_click import polyscript_tool
3. Decorate your class: @polyscript_tool
4. Implement: status(), test(), sandbox(), live() methods
5. Run: YourTool.run()

That's it! The framework does everything else.
""")

File: PolyScript/examples/backup_tool/backup_tool.go
/*
 * Example: Backup Tool using PolyScript Go Framework
 *
 * This demonstrates how the Go PolyScript framework eliminates boilerplate.
 * Shows clean Go patterns with zero CLI boilerplate.
 *
 * Author: Mathew Burkitt, Dimension Technologies <mathew.burkitt@ditech.ai>
 */

package main

import (
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"syscall"

	"github.com/spf13/cobra"
	// Import the polyscript framework
	// "path/to/polyscript"
)

// BackupTool implements the PolyScript interface
type BackupTool struct {
	// Embed DefaultTool to get default AddArguments implementation
	polyscript.DefaultTool
	sourcePath string
	destPath   string
	overwrite  bool
}

// NewBackupTool creates a new backup tool
func NewBackupTool() *BackupTool {
	return &BackupTool{
		sourcePath: "/source",
		destPath:   "/dest",
		overwrite:  false,
	}
}

func (t *BackupTool) Description() string {
	return `PolyScript-compliant backup tool with zero boilerplate

Backs up directories with full PolyScript mode support.
Provides status checking, dry-run testing, dependency validation,
and live backup operations.`
}

func (t *BackupTool) AddArguments(cmd *cobra.Command) *cobra.Command {
	cmd.Flags().String("source", "/source", "Source directory to backup")
	cmd.Flags().String("dest", "/dest", "Destination directory")
	cmd.Flags().Bool("overwrite", false, "Overwrite existing destination")
	return cmd
}

func (t *BackupTool) Status(ctx *polyscript.PolyScriptContext) (interface{}, error) {
	ctx.Log("Checking backup status...")

	sourceInfo := t.getDirectoryInfo(t.sourcePath)
	destInfo := t.getDirectoryInfo(t.destPath)

	return map[string]interface{}{
		"source": map[string]interface{}{
			"path":       t.sourcePath,
			"exists":     sourceInfo.exists,
			"size_bytes": sourceInfo.size,
			"file_count": sourceInfo.files,
		},
		"destination": map[string]interface{}{
			"path":             t.destPath,
			"exists":           destInfo.exists,
			"size_bytes":       destInfo.size,
			"would_overwrite":  destInfo.exists && !t.overwrite,
		},
		"backup_needed": sourceInfo.exists && (!destInfo.exists || t.overwrite),
	}, nil
}

func (t *BackupTool) Test(ctx *polyscript.PolyScriptContext) (interface{}, error) {
	ctx.Log("Planning backup operations...")

	sourceInfo := t.getDirectoryInfo(t.sourcePath)
	destInfo := t.getDirectoryInfo(t.destPath)

	if !sourceInfo.exists {
		ctx.Output("Source directory does not exist", true)
		return nil, polyscript.NewError("Source directory does not exist")
	}

	var operations []map[string]interface{}

	if destInfo.exists && !t.overwrite {
		operations = append(operations, map[string]interface{}{
			"operation":   "skip",
			"reason":      "destination exists and overwrite not specified",
			"source":      t.sourcePath,
			"destination": t.destPath,
		})
	} else {
		operations = append(operations, map[string]interface{}{
			"operation":      "backup",
			"source":         t.sourcePath,
			"destination":    t.destPath,
			"file_count":     sourceInfo.files,
			"size_bytes":     sourceInfo.size,
			"would_overwrite": destInfo.exists,
		})
	}

	return map[string]interface{}{
		"planned_operations": operations,
		"total_files":        sourceInfo.files,
		"total_size":         sourceInfo.size,
		"note":               "No changes made in test mode",
	}, nil
}

func (t *BackupTool) Sandbox(ctx *polyscript.PolyScriptContext) (interface{}, error) {
	ctx.Log("Testing backup environment...")

	tests := map[string]string{
		"source_readable":      t.testSourceReadable(),
		"destination_writable": t.testDestinationWritable(),
		"sufficient_space":     t.testSufficientSpace(),
		"filesystem_access":    t.testFilesystemAccess(),
	}

	allPassed := true
	for _, status := range tests {
		if status != "passed" {
			allPassed = false
			break
		}
	}

	return map[string]interface{}{
		"dependency_tests": tests,
		"all_passed":       allPassed,
	}, nil
}

func (t *BackupTool) Live(ctx *polyscript.PolyScriptContext) (interface{}, error) {
	ctx.Log("Preparing backup execution...")

	sourceInfo := t.getDirectoryInfo(t.sourcePath)
	if !sourceInfo.exists {
		ctx.Output("Source directory does not exist", true)
		return nil, polyscript.NewError("Source directory does not exist")
	}

	destInfo := t.getDirectoryInfo(t.destPath)
	if destInfo.exists && !t.overwrite {
		if !ctx.Confirm(fmt.Sprintf("Destination %s exists. Overwrite?", t.destPath)) {
			return map[string]interface{}{"status": "cancelled"}, nil
		}
	}

	ctx.Log(fmt.Sprintf("Starting backup from %s to %s", t.sourcePath, t.destPath))

	// Simulate backup operation
	// In real implementation: os.RemoveAll(t.destPath); os.Rename(t.sourcePath, t.destPath)
	// For demo, we'll just simulate
	ctx.Log("Backup operation completed (simulated)")

	resultInfo := t.getDirectoryInfo(t.destPath)

	return map[string]interface{}{
		"operation":     "backup_completed",
		"source":        t.sourcePath,
		"destination":   t.destPath,
		"files_copied":  resultInfo.files,
		"bytes_copied":  resultInfo.size,
	}, nil
}

// Helper types and methods for directory operations
type directoryInfo struct {
	exists bool
	size   int64
	files  int
}

func (t *BackupTool) getDirectoryInfo(path string) directoryInfo {
	info := directoryInfo{exists: false, size: 0, files: 0}

	stat, err := os.Stat(path)
	if err != nil || !stat.IsDir() {
		return info
	}

	info.exists = true

	// Walk directory to calculate size and count files
	filepath.WalkDir(path, func(filePath string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil // Skip errors
		}

		if !d.IsDir() {
			info.files++
			if fileInfo, err := d.Info(); err == nil {
				info.size += fileInfo.Size()
			}
		}
		return nil
	})

	return info
}

func (t *BackupTool) testSourceReadable() string {
	if stat, err := os.Stat(t.sourcePath); err == nil && stat.IsDir() {
		// Try to read directory
		if _, err := os.ReadDir(t.sourcePath); err == nil {
			return "passed"
		}
	}
	return "failed"
}

func (t *BackupTool) testDestinationWritable() string {
	destDir := filepath.Dir(t.destPath)
	if stat, err := os.Stat(destDir); err == nil && stat.IsDir() {
		// Try to create a temporary file
		tempFile := filepath.Join(destDir, ".polyscript_test")
		if file, err := os.Create(tempFile); err == nil {
			file.Close()
			os.Remove(tempFile)
			return "passed"
		}
	}
	return "failed"
}

func (t *BackupTool) testSufficientSpace() string {
	sourceInfo := t.getDirectoryInfo(t.sourcePath)
	if !sourceInfo.exists {
		return "unknown"
	}

	destDir := filepath.Dir(t.destPath)
	if stat, err := os.Stat(destDir); err == nil && stat.IsDir() {
		// Get disk space (Unix-specific)
		var statfs syscall.Statfs_t
		if syscall.Statfs(destDir, &statfs) == nil {
			available := int64(statfs.Bavail) * int64(statfs.Bsize)
			required := int64(float64(sourceInfo.size) * 1.1) // 10% buffer
			if available > required {
				return "passed"
			}
			return "failed"
		}
	}
	return "error"
}

func (t *BackupTool) testFilesystemAccess() string {
	tempFile := filepath.Join(os.TempDir(), ".polyscript_test")
	if file, err := os.Create(tempFile); err == nil {
		file.Close()
		os.Remove(tempFile)
		return "passed"
	}
	return "failed"
}

// Simple tool using minimal implementation
type SimpleTool struct {
	polyscript.DefaultTool
}

func (t *SimpleTool) Description() string {
	return "Simple backup tool demonstrating minimal Go implementation"
}

func (t *SimpleTool) Status(ctx *polyscript.PolyScriptContext) (interface{}, error) {
	ctx.Log("Simple status check...")
	return map[string]interface{}{
		"operational": true,
		"ready":       true,
	}, nil
}

func (t *SimpleTool) Test(ctx *polyscript.PolyScriptContext) (interface{}, error) {
	ctx.Log("Simple test mode...")
	return map[string]interface{}{
		"would_backup": []string{"file1", "file2"},
		"note":         "No changes made in test mode",
	}, nil
}

func (t *SimpleTool) Sandbox(ctx *polyscript.PolyScriptContext) (interface{}, error) {
	ctx.Log("Simple sandbox test...")
	return map[string]interface{}{
		"environment": "ok",
		"all_passed":  true,
	}, nil
}

func (t *SimpleTool) Live(ctx *polyscript.PolyScriptContext) (interface{}, error) {
	ctx.Log("Simple live execution...")
	if ctx.Confirm("Execute backup?") {
		return map[string]interface{}{
			"backup_completed": true,
			"files_backed_up":  42,
		}, nil
	}
	return map[string]interface{}{"status": "cancelled"}, nil
}

// Main function - choose which tool to run
func main() {
	// Uncomment the tool you want to test:
	
	// Full-featured backup tool
	tool := NewBackupTool()
	polyscript.RunTool(tool)
	
	// Simple minimal tool
	// tool := &SimpleTool{}
	// polyscript.RunTool(tool)
}

/*
 * PROJECT STRUCTURE:
 *
 * backup-tool/
 * ├── go.mod
 * ├── go.sum
 * ├── main.go (this file)
 * └── polyscript/
 *     └── framework.go
 *
 * GO.MOD:
 * module backup-tool
 * go 1.19
 * require github.com/spf13/cobra v1.7.0
 *
 * BUILD AND RUN:
 * go mod init backup-tool
 * go get github.com/spf13/cobra@latest
 * go build -o backup-tool main.go
 * ./backup-tool status
 * ./backup-tool test --verbose
 * ./backup-tool sandbox --json
 * ./backup-tool live --force
 * ./backup-tool status --source /home/user/docs --dest /backup/docs --overwrite
 *
 * The framework automatically provides:
 * - All CLI argument parsing and validation with cobra
 * - Command routing for the four PolyScript modes
 * - --json, --verbose, --force standard flags
 * - PolyScript v1.0 JSON output formatting
 * - Error handling and exit codes
 * - Help text generation
 * - Confirmation prompts
 * - Single binary distribution
 *
 * BENEFITS OF GO APPROACH:
 * - ZERO boilerplate code
 * - Single binary distribution (huge team advantage)
 * - Fast execution and startup
 * - Excellent cross-platform support
 * - Strong typing with interface contracts
 * - Simple dependency management with go mod
 * - Native performance
 * - Goroutine support for concurrent operations
 *
 * GO-SPECIFIC ADVANTAGES:
 * - Compile-time type checking
 * - No runtime dependencies
 * - Easy cross-compilation (GOOS=linux go build)
 * - Built-in testing framework
 * - Excellent tooling (go fmt, go vet, go test)
 * - Fast compilation
 * - Memory efficient
 * - Great for DevOps and system tools
 */